website/database.py

```diff
@@ -18,29 +47,93 @@
     # Production or dev: use environment variables or dev storage
     storage = dynamo.AwsDynamoStorage.from_env() or dynamo.MemoryStorage("dev_database.json")
 
-USERS = dynamo.Table(storage, "users", "username", indexes=[
-    dynamo.Index("email"),
-    dynamo.Index("epoch", sort_key="created")
-])
-TOKENS = dynamo.Table(storage, "tokens", "id", indexes=[
-    dynamo.Index('id'),
-    dynamo.Index('username'),
-])
-PROGRAMS = dynamo.Table(storage, "programs", "id", indexes=[
-    dynamo.Index('username', sort_key='date', index_name='username-index'),
-    dynamo.Index('hedy_choice', sort_key='date', index_name='hedy_choice-index'),
-    # For the explore page, this index has 'level', 'lang' and 'adventure_name'
-    dynamo.Index('public', sort_key='date'),
-
-    # For the filtered view of the 'explore' page (keys_only so we don't duplicate other attributes unnecessarily)
-    dynamo.Index('lang', sort_key='date', keys_only=True),
-    dynamo.Index('level', sort_key='date', keys_only=True),
-    dynamo.Index('adventure_name', sort_key='date', keys_only=True),
-])
-CLASSES = dynamo.Table(storage, "classes", "id", indexes=[
-    dynamo.Index('teacher'),
-    dynamo.Index('link'),
-])
+
+def only_in_dev(x):
+    """Return the argument only in debug mode. In production or offline mode, return None.
+
+    This is intended to be used with validation expressions, so that when testing
+    locally we do validation, but production data that happens to work but doesn't
+    validate doesn't throw exceptions.
+    """
+    return x if is_debug_mode() else None
+
+
+USERS = dynamo.Table(storage, 'users', 'username',
+                     types=only_in_dev({
+                         'username': str,
+                         'email': str,
+                         'epoch': int,
+                         'created': int,
+                         'classes': Optional(SetOf(str)),
+                         'keyword_language': Optional(str),
+                         'password': str,
```

Comment:
I like this!!! I think this makes a lot of sense. More typing annotations on our DB would be a huge win. I think the work now would be to adapt the definitions of all the tables. I can handle that!

Reply:
🙏 that would be great, thanks! You probably also have a better view into what the currently expected data contract is